### PR TITLE
core: Add callbacks to the `Builder`

### PR DESCRIPTION
**Pull request by math-fehr**
These callbacks are called whenever an operation or block is created and inserted.
They are used in the pattern matching engine to keep the worklist updated.

### CODE REVIEW

File: xdsl/builder.py [hunk 1]
```diff
@@ -78,6 +78,12 @@ class Builder:
     _insertion_point: InsertPoint
     """Operations will be inserted at this location."""
 
+    notify_operation_inserted: Callable[[Operation], None] | None = field(default=None)
```
Comment (superlopuh) [inline xdsl/builder.py:81]: Is there a reason you chose this name? I would have expected something like `on_operation_inserted`, or `operation_inserted_handler`. Are these the names from the cpp builder?

**Review by superlopuh — APPROVED** [review]
Code looks good small nit on naming

File: xdsl/builder.py (hunk 1 above)
Comment (PapyChacal) [reply xdsl/builder.py:81]: +1 for handler on my side
Also, why not handlers plural, with a list of callable ?

**Review by PapyChacal — APPROVED** [review]

File: xdsl/builder.py (hunk 1 above)
Comment (math-fehr) [reply xdsl/builder.py:81]: Yeah, the name come from the C++. I can change it to handler though, it's not that important to keep the same names here!
We *could* do a list, but I'm not sure if it's worth compare to nesting `Callable` together. In general, I'm never sure if callbacks should be single `Callable` or a list of `Callable`. In that particular case, I don't think we will ever remove callbacks, and there is none by default, so I would say that a single one makes sense? If you want, I could add a method `add_handler` or something, that will take care of adding one and nesting it nicely!

**Comment by math-fehr** [comment]
I fixed the naming, and replaced `callback | None` with `list[callback]`.
If that's good for you, I'll merge it!

**Review by webmiche — APPROVED** [review]
LGTM :+1:

**Review by superlopuh — APPROVED** [review]

**Pull request merged**